core/src/main/java/org/elasticsearch/search/collapse/CollapseBuilder.java

```diff
@@ -74,14 +98,30 @@ public CollapseBuilder(String field) {
     public CollapseBuilder(StreamInput in) throws IOException {
         this.field = in.readString();
         this.maxConcurrentGroupRequests = in.readVInt();
-        this.innerHit = in.readOptionalWriteable(InnerHitBuilder::new);
+        if (in.readBoolean()) {
+            int numInnerHits = in.readVInt();
+            this.innerHits = new ArrayList<>(numInnerHits);
+            for (int i = 0; i < numInnerHits; i++) {
+                innerHits.add(new InnerHitBuilder(in));
+            }
+        }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeString(field);
         out.writeVInt(maxConcurrentGroupRequests);
-        out.writeOptionalWriteable(innerHit);
+        if (innerHits != null) {
```

Comment:
Same here, you'll need to deserialize differently depending on StreamOutput#getVersion